### PR TITLE
Performance tracking

### PR DESCRIPTION
**Pull request by MajaTrzebiatowska**
Add abstraction for performance tracking along with two base rx operators (observable and single) in a base module.

Add Firebase implementation of performance tracking as a separate module.

### CODE REVIEW

File: base/src/test/java/com/mindera/skeletoid/performance/PerformanceObservableOperatorTest.kt [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        val testScheduler = TestScheduler()
+        RxJavaPlugins.setComputationSchedulerHandler { testScheduler }
+
+        Observable.create<String> { emitter ->
+            run {
+                Thread.sleep(800)
+                emitter.onComplete()
+            }
+        }
+            .subscribeOn(testScheduler)
```
Comment (bruno-garces) [inline base/src/test/java/com/mindera/skeletoid/performance/PerformanceObservableOperatorTest.kt:62]: uh, weird formatting 🙈

File: base/src/test/java/com/mindera/skeletoid/performance/PerformanceObservableOperatorTest.kt [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        verify(performanceTracker).stopTracking()
+    }
+
+    @Test
+    fun shouldStartStopPerformanceTrackingOnError() {
+        val observable: TestObserver<String> = Observable.create<String> { emitter ->
+            emitter.onError(Throwable("error"))
+        }
+            .lift(observableOperator)
+            .test()
```
Comment (bruno-garces) [inline base/src/test/java/com/mindera/skeletoid/performance/PerformanceObservableOperatorTest.kt]: same weird formatting 🙈

File: base/src/test/java/com/mindera/skeletoid/performance/PerformanceObservableOperatorTest.kt [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        verify(performanceTracker).stopTracking()
+    }
+
+    @Test
+    fun shouldNotStopPerformanceTrackingOnNext() {
+        Observable.create<String> { emitter ->
+            emitter.onNext("hello")
+        }
+            .lift(observableOperator)
+            .test()
```
Comment (bruno-garces) [inline base/src/test/java/com/mindera/skeletoid/performance/PerformanceObservableOperatorTest.kt:89]: I don't really want to spam this PR but this is a really weird formatting 🙈

File: base/src/test/java/com/mindera/skeletoid/performance/PerformanceSingleOperatorTest.kt [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        Single.create<String> { emitter ->
+            run {
+                Thread.sleep(1600)
+                emitter.onSuccess("hi")
+            }
+        }
+            .subscribeOn(testScheduler)
+            .observeOn(testScheduler)
+            .lift(operator)
+            .test()
```
Comment (bruno-garces) [inline base/src/test/java/com/mindera/skeletoid/performance/PerformanceSingleOperatorTest.kt:65]: here we go again 🙈

File: base/src/test/java/com/mindera/skeletoid/performance/PerformanceSingleOperatorTest.kt [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        verify(performanceTracker).stopTracking()
+    }
+
+    @Test
+    fun shouldStartStopPerformanceTrackingOnError() {
+        val observable: TestObserver<String> = Single.create<String> { emitter ->
+            emitter.onError(Throwable("error"))
+        }
+            .lift(operator)
+            .test()
```
Comment (bruno-garces) [inline base/src/test/java/com/mindera/skeletoid/performance/PerformanceSingleOperatorTest.kt]: same 🙈

**Review by bruno-garces — APPROVED** [review]

**Review by neteinstein — APPROVED** [review]

**Review by djrsousa — APPROVED** [review]

File: base/src/test/java/com/mindera/skeletoid/performance/PerformanceObservableOperatorTest.kt (hunk 1 above)
Comment (MajaTrzebiatowska) [reply base/src/test/java/com/mindera/skeletoid/performance/PerformanceObservableOperatorTest.kt:62]: what do you suggest? whatever I try doing, AS keeps changing it when I use autoformat :D

Comment (bruno-garces) [reply base/src/test/java/com/mindera/skeletoid/performance/PerformanceObservableOperatorTest.kt:62]: Yeah, I know. It's one of those cases where we can't rely on AS 🤷‍♂️

Comment (MajaTrzebiatowska) [reply base/src/test/java/com/mindera/skeletoid/performance/PerformanceObservableOperatorTest.kt:62]: I'd say leave as is for now, as it'll change every time someone opens the files and decides to format the whole class. We can revisit once we have new coded formatting style guides.

**Pull request merged**